### ISSUE TITLE
RigidBody2d not respecting changes in position triggered by signals

### ISSUE THREAD

**Issue by Proggle** [godotengine/godot#74323]
### Godot version

v4.0.stable.official [92bee43ad]

### System information

Window 10

### Issue description

Here's a rigidBody2d ball.  Every 4 bounces it's supposed to return to the starting position.  As you can see, it does, but only for 1 frame.
https://user-images.githubusercontent.com/89282845/222824775-067b4a20-a931-4b87-909d-409cbf9aa917.mp4

Basically, if you use signals to change positions of a rigidbody2d, it doesn't work; the object teleports to the correct position for 1 frame, then teleports back.  While the documentation says you shouldn't change the position of a Rigidbody2D 'every frame' or 'very often', it seems like if this behavior is working as intended, the documentation should instead say 'you cannot directly change the position under any circumstances'.

In principle it's possible to work around this (by caching your new position or velocity or whatever and putting it into the rigid body physics _integrate_forces function), but that makes working with signals a lot more cumbersome.  It would be a lot better if godot handled this gracefully, or (failing that) if there was a built-in cache variable (self.nextTickPosition?) or function (self.moveTo()?) to make it move to the right location on the next physics calculation.  No need to have everyone reimplement them on every rigidBody2d in their project.

### Steps to reproduce

Change the position of a rigidbody2d in a signal.  It will move to that position for exactly one frame, then return to where it was.

### Minimal reproduction project

[ballTeleportBug.zip](https://github.com/godotengine/godot/files/10885887/ballTeleportBug.zip)

Here's a minimal project.  Every 4 bounces it's supposed to return to the starting position.  As you can see, it does, but only for 1 frame.

**Comment by Proggle** [comment]
Followup: the documentation says that changing transform will never work
>Please also keep in mind that physics bodies manage their own transform which overwrites the ones you set. So any direct or indirect transformation (including scaling of the node or its parent) will be visible in the editor only, and immediately reset at runtime.

But replacing self.global_position = newPosition with self.global_transform.origin = newPosition works perfectly fine, without bugs.

Maybe this is just a documentation issue?

**Comment by Calinou** [comment]
@Proggle Remember that for GitHub video previews to work, you need to have a blank line before *and* after the video URL. I edited your post accordingly, but remember to do this in the future :slightly_smiling_face:

**Comment by sweti-yeti** [comment]
I don't think this is strictly related to setting the transform via a signal -- I'm running into the same issue when setting the transform via _integrate_forces. My code is something like:

```
extends RigidBody2D

@onready var start_transform := self.transform
var should_reset = false

func _ready():
		Signals.reset_rigidbody.connect(_on_reset)

func _integrate_forces(state: PhysicsDirectBodyState2D):
	if should_reset:
		state.transform = start_transform
		should_reset = false

func _on_reset():
	should_reset = true
```

And I see the same behavior of appearing in the starting position for only 1 frame.

**Comment by sweti-yeti** [comment]
So, after experimenting with the minimal repro posted in the issue: The problem here is specifically a difference in behavior between `transform` and `global_transform`. I tried setting transform, global_transform, position, and global_position via the `_integrate_forces` and `_physics_process` functions, and the only one that worked was setting `global_transform` in `_physics_process`. All other combinations exhibited the "teleporting for a single frame" behavior. I would at least expect `transform` and `global_transform` to behave consistently (aside from the obvious global vs local difference)

I can expand the example project to show the differences and either attach it here or create a new issue, if that would be helpful.

**Comment by Proggle** [comment]
> So, after experimenting with the minimal repro posted in the issue: The problem here is specifically a difference in behavior between `transform` and `global_transform`. I tried setting transform, global_transform, position, and global_position via the `_integrate_forces` and `_physics_process` functions, and the only one that worked was setting `global_transform` in `_physics_process`. All other combinations exhibited the "teleporting for a single frame" behavior. I would at least expect `transform` and `global_transform` to behave consistently (aside from the obvious global vs local difference)
> 
> I can expand the example project to show the differences and either attach it here or create a new issue, if that would be helpful.

I'd appreciate that - it's pretty clear that this bug is bigger than what I was initially thinking.

**Comment by Saltlight-Studio** [comment]
Hello, wanted to mention that I found this while troubleshooting an odd Issue I had with Rigidbody2D too, and self.global_transform.origin = node.global_position fixed my issue.

In my case I had a Pinjoint2D attached to a CharacterBody2D and said Rigidbody2D.

I needed to freeze and unfreeze my RigidBody2D periodically while resetting its position.

Before using transform, when having set a new global position when I un-freezed it, it would teleport back to where it was last frozen only to then slingshot to the new position.

If it helps I noticed that setting center_of_mass to be equal to the global_position, it would give me this error:
"slime_body.gd:47 @ attach_to_slot(): Condition "center_of_mass_mode != CENTER_OF_MASS_MODE_CUSTOM" is true."

But it would act as I intended it to. If I hovever set the center of mass mode to Custom, the slingshot effect would return with the Rigidbody now getting stuck in certain rotations.

This was the slingshotting behaviour I got:

https://github.com/godotengine/godot/assets/47177904/137e12d5-9c0c-4af7-874f-ac58137f12c8

**Comment by Saltlight-Studio** [comment]
slime_unit.global_position = slot.global_position
self.global_position = slime_unit.global_position    <- This is what I had to change out for it to start working
self.global_rotation = 0.0
self.freeze = true

And the slingshotting happened as soon as I set self.freeze = false again, regardless of other code

**Comment by Saltlight-Studio** [comment]
Oh and I am also making use of signals, but I am not sending the position vectors, just calling the signal on a timer delay for the reattachment and freezing to happen

**Comment by Zifai** [comment]
> Followup: the documentation says that changing transform will never work
> 
> > Please also keep in mind that physics bodies manage their own transform which overwrites the ones you set. So any direct or indirect transformation (including scaling of the node or its parent) will be visible in the editor only, and immediately reset at runtime.
> 
> But replacing self.global_position = newPosition with self.global_transform.origin = newPosition works perfectly fine, without bugs.
> 
> Maybe this is mostly a documentation issue? It's weird that position and transform.origin are different, though.

is working
